### PR TITLE
SASL: Add support for EXTERNAL method.

### PR DESCRIPTION
**Pull request by IceN9ne**
- Adds dynamic fallback from EXTERNAL to PLAIN if auth fails on the network.

### CODE REVIEW

**Comment by AndrioCelos** [comment]
Seems to be working well. It may be worth adding a GUI warning or error message if you select SASL EXTERNAL with no SSL certificate or without enabling SSL.

**Comment by DarthGandalf** [comment]
The following is the kind of fallback I had in mind...

C: AUTHENTICATE EXTERNAL
S: AUTHENTICATE +
C: AUTHENTICATE +
S: :server 904 nick :SASL authentication failed
~~C: aaaaaa, I failed! what to do? disconnect!~~
C: AUTHENTICATE PLAIN
S: AUTHENTCIATE +
...

File: src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp [hunk 1]
```diff
@@ -3039,6 +3039,30 @@ void KviIrcServerParser::parseNumericSaslFail(KviIrcMessage * msg)
...
 
+	// Handle fallback if possible for SASL auth failure or dump if user mandates SASL
+	// and no fallback is available
+	if(msg->numeric() == 904)
+	{
+		if(msg->connection()->stateData()->sentSaslMethod() == QStringLiteral("EXTERNAL"))
+		{
+			if(!msg->connection()->target()->server()->saslNick().isEmpty() && !msg->connection()->target()->server()->saslPass().isEmpty())
+			{
+				KviWindow * pOut = static_cast<KviWindow *>(msg->console());
```
Comment (DarthGandalf) [inline src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp:3050]: It doesn't work without static_cast?

File: src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp [hunk 2]
```diff
@@ -3039,6 +3039,30 @@ void KviIrcServerParser::parseNumericSaslFail(KviIrcMessage * msg)
 		pOut->output(KVI_OUT_SERVERINFO, __tr2qs("SASL authentication error: %Q"), &szParam);
 	}
 
+	// Handle fallback if possible for SASL auth failure or dump if user mandates SASL
+	// and no fallback is available
+	if(msg->numeric() == 904)
+	{
+		if(msg->connection()->stateData()->sentSaslMethod() == QStringLiteral("EXTERNAL"))
+		{
+			if(!msg->connection()->target()->server()->saslNick().isEmpty() && !msg->connection()->target()->server()->saslPass().isEmpty())
```
Comment (DarthGandalf) [inline src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp:3048]: I'd probably refactor this to not do the same several times...

File: src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp [hunk 3]
```diff
@@ -3039,6 +3039,30 @@ void KviIrcServerParser::parseNumericSaslFail(KviIrcMessage * msg)
...
+				msg->connection()->stateData()->setSentSaslMethod(QStringLiteral("PLAIN"));
+				return;
+			}
+		}
+
+		if(KVI_OPTION_BOOL(KviOption_boolDropConnectionOnSaslFailure))
+		{
+			KviWindow * pOut = static_cast<KviWindow *>(msg->console());
+			pOut->output(KVI_OUT_SERVERINFO, __tr2qs("SASL auth failed. Dropping the connection."));
+			msg->connection()->sendFmtData("QUIT");
```
Comment (DarthGandalf) [inline src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp]: Server isn't guaranteed to disconect on QUIT though, even if it probably will... Can we send QUIT and then disconnect?

File: src/modules/options/OptionsWidget_connection.cpp [hunk 4]
```diff
@@ -182,7 +182,10 @@ OptionsWidget_connectionSocket::OptionsWidget_connectionSocket(QWidget * parent)
 	                        "you want to rely on the DNS server to provide the best choice.",
 	                "options"));
 
-	addRowSpacer(0, 5, 0, 5);
+	b = addBoolSelector(0, 5, 0, 5, __tr2qs_ctx("Drop connection on SASL authentication failure", "options"), KviOption_boolDropConnectionOnSaslFailure);
```
Comment (DarthGandalf) [inline src/modules/options/OptionsWidget_connection.cpp]: Should be in a separate commit

**Review by DarthGandalf — APPROVED** [review]
If you tested the fallback, LGTM.

Also it would be good to parse 908 before attempting the fallback, but not necessary.

File: src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp (hunk 1 above)
Comment (IceN9ne) [reply src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp:3050]: I literally just copied that from the code that was already there to keep it consistent.

File: src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp (hunk 2 above)
Comment (IceN9ne) [reply src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp:3048]: Do what multiple times?

File: src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp (hunk 3 above)
Comment (IceN9ne) [reply src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp]: The idea is to not force a disconnect so that it will automatically attempt to connect again, in case services are having issues, but I suppose in most situations where services would be having issues, you wouldn't get a 904 response.

Comment (DarthGandalf) [reply src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp]: I think it depends on server implementation. If server is evil (it's remote untrusted input after all), we don't want it to trick us into continuing connection even if user told us to disconnect in this case.

File: src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp (hunk 2 above)
Comment (DarthGandalf) [reply src/kvirc/sparser/KviIrcServerParser_numericHandlers.cpp:3048]: Have several pieces of code which check whether PLAIN is valid.
If one will need to be changed later, the other one will be forgotten.
